[PersistentStorge.FileSystem] Readme update (#1802)
open-telemetry/opentelemetry-dotnet-contrib: src/OpenTelemetry.PersistentStorage.FileSystem/README.md

dotnet add package OpenTelemetry.PersistentStorage.FileSystem
```

## Usage

### Setup FileBlobProvider

```csharp

> using var persistentBlobProvider = new FileBlobProvider("@C:\temp");

Reviewer: ```suggestion
using var persistentBlobProvider = new FileBlobProvider(@"C:\temp");
```

```

#### Configuration

> Following is the complete list of configurable options that can be used to set
> up FileBlobProvider:

Reviewer: ```suggestion
Following is the complete list of parameters the `FileBlobProvider` constructor
accepts to control its configuration:
```

Kind of a mouthful but I wanted to point out they are ctor parameters. It wasn't clear to me at first because they were being called "options" which is usually a class.

##### path

> Sets folder location where blobs are stored.

Reviewer: ```suggestion
Sets folder location where blobs are stored. Required.
```

I'm not going to drop comments on all the other things, but maybe add "Optional" to them? Like...

```
##### maxSizeInBytes

Maximum allowed folder size. Optional. Default if not specified: `50 MB`.
```

Author: updated

##### maxSizeInBytes

> Maximum allowed folder size. Default is `50 MB`.

Reviewer: What happens when the max is reached? New blobs/files are dropped/ignored? Is this logged anywhere?

Author: New blobs are ignored. We do log it.

##### maintenancePeriodInMilliseconds

Maintenance event runs at specified interval. Default is `2 minutes`. During
this event, the following tasks are performed:

> * Removes `*.blob` files for which the retention period has expired.
> * Removes `*.tmp` files for which the write timeout period has expired.
> * Updates `*.lock` files to `*.blob` for which the lease period has expired.
> * Updates `directorySize`.

Reviewer: nit: Seems to read more nicely to me with singular form:

```suggestion
* Remove `*.blob` files for which the retention period has expired.
* Remove `*.tmp` files for which the write timeout period has expired.
* Update `*.lock` files to `*.blob` for which the lease period has expired.
* Update `directorySize`.
```

dotnet add package OpenTelemetry.PersistentStorage.FileSystem
```

## Usage

### Setup FileBlobProvider

```csharp
using var persistentBlobProvider = new FileBlobProvider(@"C:\temp");
```

#### Configuration

Following is the complete list of parameters that `FileBlobProvider` constructor
accepts to control its configuration:

##### path

> Sets folder location where blobs are stored. `Required`.

Reviewer 2: nit: 
```suggestion
Sets directory location where blobs are stored. `Required`.
```

##### maxSizeInBytes

Maximum allowed folder size. `Optional`. Default if not specified: `52428800`
bytes.

New blobs are dropped after the folder size reaches maximum limit.

##### maintenancePeriodInMilliseconds

Maintenance event runs at specified interval. `Optional`. Default if not
specified: `120000`ms.

During this event, the following tasks are performed:

> * Remove `*.blob` files for which the retention period has expired.

Reviewer 2: Do you want to call out these time periods are stamped on the file. (or) Provide a link to file naming below and speak about how the time is utilized.

Author: added a note below with link.

dotnet add package OpenTelemetry.PersistentStorage.FileSystem
```

## Usage

### Setup FileBlobProvider

```csharp
using var persistentBlobProvider = new FileBlobProvider(@"C:\temp");
```

#### Configuration

Following is the complete list of parameters that `FileBlobProvider` constructor
accepts to control its configuration:

##### path

Sets directory location where blobs are stored. `Required`.

##### maxSizeInBytes

Maximum allowed folder size. `Optional`. Default if not specified: `52428800`
bytes.

> New blobs are dropped after the folder size reaches maximum limit.

Reviewer: I think we should do something like...

```
New blobs are dropped after the folder size reaches maximum limit. A log message is written if blobs cannot be written. See [Troubleshooting](#troubleshooting) for more information.
```

Then add a section a the bottom for "Troubleshooting" which mentions the event source and links to self-diagnostics feature for how to capture that info?

Author: added

    }
}
```

## File Details

### File naming

Each call to [CreateBlob](#createblob) methods create a single file(blob) at the
configured [directory path](#path). Each file that is created has unique
name in the format `yyyy-MM-ddTHHmmss.fffffffZ-GUID`. The file extension depends
on the operation. When creating a blob, the file is stored with the `.blob`
extension. If a lease is acquired on an existing file or on the file being
created, the file extension is changed to `.lock`, along with the lease
expiration time appended to its name in the format
`@yyyy-MM-ddTHHmmss.fffffffZ`. The `.tmp` extension is used for files while data

> writing is in process.

Reviewer 2: Provide an example file name for all 3 extensions below.

### File Access

Access to the stored files is controlled by the underlying file system
permissions set on the directory where the files are stored.

### Data format

The data contained within the file is stored in its original, unprocessed format
provided in the byte array.

### Data retention

A blob stored on disk persists until it is explicitly deleted using the
[TryDelete](#delete) operation or is removed during the [maintenance
job](#maintenanceperiodinmilliseconds) upon the expiration of its

> [retention](#retentionperiodinmilliseconds) period.

Reviewer 2: Require a section on data security. You could refer this document - https://learn.microsoft.com/en-us/azure/azure-monitor/app/telemetry-channels#does-the-sdk-create-temporary-local-storage-is-the-data-encrypted-at-storage

Author: I changed the data format part to include this. does that look good to you?

Author: added

    }
}
```

## File Details

### File naming

Each call to [CreateBlob](#createblob) methods create a single file(blob) at the
configured [directory path](#path). Each file that is created has unique name in

> the format `yyyy-MM-ddTHHmmss.fffffffZ-GUID`. The file extension depends on the

Reviewer 2: Explain what time is used here, is it in utc or other time zone.

Reviewer 2: Same applies for lease.

operation. When creating a blob, the file is stored with the `.blob`

> extension(e.g. `2024-05-15T174825.3027972Z-40386ee02b8a47f1b04afc281f33d712.blob

Reviewer 2: You could move all the examples to the end of the section. Something like below.

Example files: 

```
2024-05-15T174825.3027972Z-40386ee02b8a47f1b04afc281f33d712.blob
2024-05-15T174825.3027972Z-40386ee02b8a47f1b04afc281f33d712.blob@2024-05-15T203551.2932278Z.lock
2024-05-15T175941.2228167Z-6649ff8ce55144b88a99c440a0b9feea.blob.tmp
```

    }
}
```

## File Details

### File naming

Each call to [CreateBlob](#createblob) methods create a single file(blob) at the
configured [directory path](#path). Each file that is created has unique name in

> the format `datetimestamp(ISO 8601)-GUID`. The file extension depends on the

Reviewer 2: ```suggestion
the format `datetimestamp(ISO 8601)-GUID` with current datetime. The file extension depends on the
```

operation. When creating a blob, the file is stored with the `.blob` extension.
If a lease is acquired on an existing file or on the file being created, the
file extension is changed to `.lock`, along with the lease expiration time
appended to its name in the format `@datetimestamp(ISO 8601)`. The `.tmp`
extension is used for files while data writing is in process.

Example file names:

* `2024-05-15T174825.3027972Z-40386ee02b8a47f1b04afc281f33d712.blob`
* `2024-05-15T174825.3027972Z-40386ee02b8a47f1b04afc281f33d712.blob@2024-05-15T203551.2932278Z.lock`
* `2024-05-15T175941.2228167Z-6649ff8ce55144b88a99c440a0b9feea.blob.tmp`

### Data format and security

The data contained within the file(blob) is unencrypted and stored in its

> original, unprocessed format provided in the byte array. If specific security

Reviewer 3: privacy or security

Author: This is security.

> requirements need to be met, application owners MUST configure the

Reviewer 3: Maybe SHOULD?
